content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+
+```java
+@Controller
+public class UserController() {
+    @Autowired
+    private UserMapper userMapper;
+}
+```
+
+Similarly, if we are not using Spring framework, MapStruct has the support for [CDI](https://docs.oracle.com/javaee/6/tutorial/doc/giwhl.html) as well:
+
+```java
+@Mapper(componentModel = "cdi", uses = {ManagerMapper.class, Validator.class})
+public interface UserMapper {
+    
+    ...
+}
+```
+
+Then the generated mapper implementation will be annotated with `@ApplicationScoped` annotation:
+
+```java
+@ApplicationScoped
+public class UserMapperImpl implements UserMapper {
+	...
+}
+```
+
+Finally, we can obtain the constructor using the `@Inject` annotation:
+
+```java
+@Inject
+private UserMapper userMapper;
+```
+
+## Data Type Conversion
+
+We won’t always find a mapping attribute in a payload having the same data type for the source and target fields. For example, we might have an instance where we would need to map an attribute of type `int` to `String` or `long`. We will take a quick look at how we can deal with such types of data conversions.
+
+### Implicit Type Conversion
+
+MapStruct explicitly takes care of type conversion automatically. If we want to transform an attribute of type `int` to `String`, then MapStruct converts it without defining anything in the mapper interface using `String.valueOf(int)` method. Similarly, it takes care of type conversion between primitive and wrapper types, enum types, and String, or `BigDecimal/BigInteger` type to `String`.
+
+Sometimes we would like to define a particular format while converting between certain types. For example, let us say we want to convert the field `salary` from `long` to `String` in the decimal form of US dollars:
+
+```java
+@Mapping(source = "employment.salary",
+         target = "salary",
+         numberFormat = "$#.00")
+PersonDTO convert(BasicUser user,
+                  Education education,
+                  Address address,
+                  Employment employment);
+```
+
+Then the generated mapper implementation class would be something like below:
+
+```java
+personDTO.setSalary( new DecimalFormat( "$#.00" ).format(
+                employment.getSalary() ) );
+```
+
+Similarly, let’s say if we want to convert a date type in `String` format to `LocalDate` format, then we can define a mapper in the following format:
+
+```java
+@Mapping(source = "dateOfBirth",
+         target = "dateOfBirth",
+         dateFormat = "dd/MMM/yyyy")
+ManagerDTO convert(Manager manager);
+```
+
+Then the generated mapper implementation would be something like below:
+
+```java
+managerDTO.setDateOfBirth(
+    new SimpleDateFormat( "dd/MMM/yyyy" )
+    .parse( manager.getDateOfBirth() ) );
+```
+
+If we don’t mention the `dateFormat` property in above mapper then this would generate an implementation method something like below:
+
+```java
+managerDTO.setDateOfBirth( new SimpleDateFormat().parse(
+    manager.getDateOfBirth() ) );
+```
+
+### Mapping Collections
```

Comment:
Does it make sense that this and the following sections are all subsections of the "Data Type Conversion" section?

Reply:
Yes, all of these sub-sections are focussed to show conversion from one data type to other, such as one collection to other or Enum to strings or vice-versa. Hence, accumulated all of them into this section.